File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.c


int maxProfit(int* prices, int pricesSize){
    if(pricesSize==0||pricesSize==1){
        return 0;
    }
    int buy=prices[0],profit=0;
   int i=1;
  while(i<pricesSize){
          if(profit<prices[i]-buy){
        profit=prices[i]-buy;  
    }
          
        if(prices[i]<buy){
            buy=prices[i];  
        }
      
    
i++;
}
    
    
        return profit;
}

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int n =prices.length;
        int[] sell = new int[n];
        int max=prices[n-1];
        sell[n-1]=prices[n-1];
        for(int i=n-2;i>=0;i--) {
              max=Math.max(sell[i+1],prices[i]);
              sell[i]=max;
        }
        int min=prices[0];
        int res=0;
        for(int i=1;i<n;i++){
                res=Math.max(res,sell[i]-min);
                min=Math.min(min,prices[i]);
                
        }
         return res;
    }
}